### ISSUE TITLE
Progress Bar

### ISSUE THREAD

**Issue by J0K3RAS** [pytube/pytube#169]
I am writing an algorithm using kivy and pytube library, according to which the user will paste the link of a youtube video and after selecting a stream, the download will begin by clicking the download button. During the downloading process, I want to add a progress bar. I have read documentation but I need some further explanation. The simplified code to trigger the problem I am dealing with is this:

```
from pytube import YouTube
from kivy.uix.progressbar import ProgressBar
from kivy.app import App
from kivy.uix.button import Button

class TestApp(App):

    def build(self):
        song = YouTube('https://www.youtube.com/watch?v=--tFFz44zvc')
        song = song.streams.get_by_itag(251)
        return song.register_on_progress_callback(self.show_progress_bar)
    
    def show_progress_bar(self, stream, chunk, file_handle, bytes_remaining):
        pb = ProgressBar(max=100)
        while bytes_remaining==0:
            downloaded = chunk - bytes_remaining
            percentage = (int(downloaded)/int(chunk))*100
            pb.value=percentage


if __name__ == '__main__':
    TestApp().run()
```

the output is : 

```
File "/home/charalampos/Desktop/test Progress Bar.py", line 25, in build
    return song.register_on_progress_callback(self.show_progress_bar)

AttributeError: 'Stream' object has no attribute 'register_on_progress_callback'
```

**Comment by nficano** [comment]
you're clobbering your ``song`` variable, what you want to do is:

```python
    def build(self):
        yt = YouTube('https://www.youtube.com/watch?v=--tFFz44zvc')
        yt.register_on_progress_callback(self.show_progress_bar)
        song = yt.streams.get_by_itag(251)
        song.download()
```

**Comment by J0K3RAS** [comment]
Thank you for response. I did the change you suggested but now the output I get is this error:

```
  File "/usr/lib/python3.6/site-packages/pytube/__main__.py", line 174, in register_on_progress_callback
    self._monostate['on_progress'] = func

AttributeError: 'YouTube' object has no attribute '_monostate'  
```

**Comment by nficano** [comment]
AH looks like I introduced a regression the other night. that'll teach me to release updates on the subway :) I just pushed a fix, please run: pip install -U pytube

**Comment by nficano** [comment]
also, it may be helpful to checkout the CLI: https://github.com/nficano/pytube/blob/master/pytube/cli.py#L162-L163

**Comment by J0K3RAS** [comment]
Now the progress bar is working but it updates once the download it completed. I guess that has to do with my code. 

Thanks for your help :)

**Comment by J0K3RAS** [comment]
I found some spare time to search a bit more about why my progress bar is updating only when the downloading process is completed. The thing is that yt.register_on_progress_callback(self.show_progress_bar) keeps running continuously, which makes kivy's GUI freeze until the process has finished. Is there any way to make yt.register_on_progress_callback(self.show_progress_bar) run only once in x seconds?

My code:
```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 23 21:18:14 2017

@author: charalampos
"""

from pytube import YouTube

from kivy.uix.progressbar import ProgressBar
from kivy.app import App
from kivy.uix.button import Button
from kivy.uix.boxlayout import BoxLayout
from threading import Thread

class TestApp(App):

    def build(self):
        global pb
        layout = BoxLayout(padding=10, orientation='vertical')
        button = Button(text='download')
        button.bind(on_release=self.ButtonClicked)
        layout.add_widget(button)
        pb = ProgressBar(max=100)
        pb.value = 0
        layout.add_widget(pb)
        return layout
    
    def ButtonClicked(self, button):
        global yt
        yt = YouTube('https://www.youtube.com/watch?v=nPJdc8qWp48')
        yt.register_on_progress_callback(self.show_progress_bar)
        song = yt.streams.get_by_itag(251)
        Thread(target = song.download).start()
              
    def show_progress_bar(self, stream, chunk, file_handle, bytes_remaining):
            l = [x for x in range(101)]
            total = stream.filesize
            if bytes_remaining==0:
                downloaded = total - bytes_remaining
                percentage = ((downloaded)/(float(total)))*100
                if float(percentage) in l:
                    pb.value = percentage
    
if __name__ == '__main__':
    TestApp().run()
```

**Comment by nficano** [comment]
it's easiest to reference how the cli progress bar works, but off the top of my head:

```python
# this needs to be removed
if bytes_remaining==0:
```

**Comment by J0K3RAS** [comment]
You are right that it is incorrect `if bytes_remaining==0:` . Actually that was a typing mistake that I did because if you check the code I posted at the beginning it was `while bytes_remaining==0:`

I read the CLI but the problem I have is related with kivy. The GUI is not being updated until the downloading process is  completed. That's why I want to add a small break (1/60 seconds) between each loop of the `yt.register_on_progress_callback(self.show_progress_bar) `  in order to give kivy the ability to update the GUI (My progress bar is part of it). I have tried `Thread(target=yt.register_on_progress_callback(self.show_progress_bar)).start()` with no success. 
The result I get in any case is my progress updating after the downloading process is completed.

This describes my problem: https://groups.google.com/forum/#!topic/kivy-users/CmeHmuwhplo

But the guy there solved the problem by adding ` Clock.schedule_interval(self.heavyFunc,1/60)`
In my case `Clock.schedule_interval(yt.register_on_progress_callback(self.show_progress_bar),1/60)` returns error because `yt.register_on_progress_callback(self.show_progress_bar)` is `None Type`
